L26-Ability-to-restrict-max-threads-in-C++SyncServer.md

```diff
@@ -0,0 +1,93 @@
+* Author(s): Sree Kuchibhotla (sreecha)
+* Approver: vjpai
+* Status: In Review
+* Implemented in: C++
+* Last updated: June 6, 2018
+* Discussion at: 
+
+## Abstract
+C++ `ThreadManager` is a specialized thread pool used to implement the C++ Synchronous server. 
+
+Currently, `ThreadManager` can be configured to have minimum and maximum number of polling threads. While this helps to always have threads available to poll for work, it does nothing to throttle polling if the `ThreadManager` is overloaded with work.
+
+The proposal here is to add a notion of 'thread quota' to the `resource_quota` object. Currently, a resource quota object can be attached to the server (via `ServerBuilder::SetResourceQuota`). The idea here is to set a maximum number of threads on that resource quota object.
+
+Each of the `ThreadManager` objects in a given server check with the Server's resource quota object when creating new threads and will not create new threads if no quota is available.
+
+## Details
+More concretely, the following API changes are being proposed
+
+### 1. New *public* C++ API: `ServerBuilder::SetMaxThreads`
+ 
+```C++
+// File: include/grpcpp/resource_quota.h
+
+/// ResourceQuota represents a bound on memory and threads usage by the gRPC
+/// library. A ResourceQuota can be attached to a server (via \a ServerBuilder),
+/// or a client channel (via \a ChannelArguments).
+///
+/// gRPC will attempt to keep memory and threads used by all attached entities
+/// below the ResourceQuota bound.
+class ResourceQuota final : private GrpcLibraryCodegen {
+ public:
+  ...
+  /// Set the max number of threads that can be allocated from this
+  /// ResourceQuota object.
+  ///
+  /// If the new_max_threads value is smaller than the current value, no new
+  /// threads are allocated until the number of active threads fall below
+  /// new_max_threads. There is no time bound on when this may happen i.e none
+  /// of the current threads are forcefully destroyed and all threads run their
+  /// normal course.
+  ResourceQuota& SetMaxThreads(int new_max_threads);
+  ...     
+```
+* Max threads are set to INT_MAX by default
+* In the initial implementation, I plan to divide the max_threads equally among multiple `ThreadManager` objects (there are as many `ThreadManager` objects as the number of server completion queues). It is not clear on which strategy is better at this point 
+  - (1) Have all thread managers create threads from a common pool (but potentially starving some thread managers and also making the quota-check a potential global contention point)
```

Comment:
Is thread creation a common enough occurrence that there's worry about contention?

Reply:
It could be if there are multiple server completion queues (and even though right now a Sync server creates just one server completion queue, in future we probably want to increase it to be equal to the number of cores).

FWIW, since we only have one server completion queue currently, both the strategies above will effectively be the same..